File: endpoints/ocaml/example_sync.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* Copyright 2026 ZeroDDS Contributors *)
(*
   Deep example (sync): a realistic sensor-telemetry flow. A publisher frames
   five typed `Reading { id; value; label }` samples and delivers them; the
   subscriber owns the run-loop and polls, decoding EVERY field byte-for-byte. *)

module Reading = struct
  type t = { id : int; value : float; label : string }

  let marshal (v : t) (endian : Zerodds.Wire.endian) : bytes =
    let open Zerodds.Wire in
    let w = writer endian in
    put_u32 w v.id;
    put_f32 w v.value;
    put_string w v.label;
    bytes w

  let decode (body : bytes) : t =
    let open Zerodds.Wire in
    let r = reader body LE in
    let id = get_u32 r in
    let value = get_f32 r in
    let label = get_string r in
    { id; value; label }
end

let () =
  let total = 5 in
  let t = Zerodds.MemTransport.create () in
  let c = Zerodds.Client.create t in
  for i = 0 to total - 1 do
    let r =
      Reading.
        {
          id = 0x1000 + i;
          value = 20.0 +. (float_of_int i *. 0.5);
          label = Printf.sprintf "bay-%02d" i;
        }
    in
    Zerodds.Client.write c (Reading.marshal r Zerodds.Wire.LE)
  done;
  let got = ref 0 in
  let continue = ref true in
  while !got < total && !continue do
    match Zerodds.Client.poll c with
    | Some body ->
        let { Reading.id; value; label } = Reading.decode body in
        Printf.printf "sync reading %d: id=0x%x value=%.1f label=\"%s\"\n" !got id
          value label;
        incr got
    | None -> continue := false
  done;
  if !got <> total then (
    Printf.eprintf "incomplete: got %d of %d\n" !got total;
    exit 1);
  print_string "ALL OK\n"

File: endpoints/java/ExampleSync.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Deep example (sync): a realistic sensor-telemetry flow. A publisher frames
// five typed Reading(id, value, label) samples and delivers them; the
// subscriber owns the run-loop and polls, decoding EVERY field byte-for-byte.

import java.util.ArrayDeque;

public final class ExampleSync {

    // Reading { uint32 id; float value; string label }.
    static final class Reading {
        final long   id;
        final float  value;
        final String label;

        Reading(long id, float value, String label) {
            this.id = id;
            this.value = value;
            this.label = label;
        }

        byte[] marshal(int endian) {
            Zdw.Writer w = new Zdw.Writer(endian);
            w.u32(id);
            w.f32(value);
            w.str(label);
            return w.bytes();
        }

        static Reading decode(byte[] body) {
            Zdw.Reader r = new Zdw.Reader(body, Zdw.LE);
            return new Reading(r.u32(), r.f32(), r.str());
        }
    }

    public static void main(String[] args) {
        final int total = 5;
        ArrayDeque<byte[]> transport = new ArrayDeque<byte[]>();

        for (int i = 0; i < total; i++) {
            Reading rd = new Reading(0x1000L + i, 20.0f + i * 0.5f,
                                     String.format("bay-%02d", i));
            transport.add(ZdwEndpoint.xrceWriteFrame(ZdwEndpoint.SESSION_NOKEY,
                ZdwEndpoint.STREAM_BEST_EFFORT, i + 1, rd.marshal(Zdw.LE)));
        }

        int got = 0;
        while (got < total) {
            byte[] frame = transport.poll();
            if (frame == null) {
                break;
            }
            Reading rd = Reading.decode(ZdwEndpoint.xrceReadBody(frame));
            System.out.printf("sync reading %d: id=0x%x value=%.1f label=\"%s\"%n",
                              got, rd.id, rd.value, rd.label);
            got++;
        }

        if (got != total) {
            System.err.println("incomplete: got " + got + " of " + total);
            System.exit(1);
        }
        System.out.println("ALL OK");
    }
}
